graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightResource.java

```diff
@@ -66,30 +71,30 @@ public List<Node> listDataNodes() {
     @GET
     @Path("/ca")
     public CA get() {
-        return currentCA; //TODO
+        return caService.get();
     }
 
     @POST
     @Path("/ca/create")
     @NoAuditEvent("No Audit Event needed")
-    public void createCA() {
-        currentCA = new CA("generated CA", CAType.GENERATED); //TODO
+    public void createCA(@FormParam("password") String password) throws CACreationException {
+        // TODO: get validity from preflight UI
+        caService.create(password, null);
     }
 
     @POST
     @Path("/ca/upload")
     @NoAuditEvent("No Audit Event needed")
     @Consumes(MediaType.APPLICATION_JSON)
-    public void uploadCA() { //TODO: input
-        currentCA = new CA("uploaded CA", CAType.UPLOADED); //TODO
+    public void uploadCA(@FormParam("password") String password, @FormParam("files") List<String> caFiles) throws CACreationException {
```

Comment:
That's probably a sad thing about the whole upload concept - password for CA files is to be inserted in the UI, which people may consider insecure...

Reply:
yes, but we use it as a one time password only as discussed